whisperv6/whisper.go

```diff
@@ -843,12 +845,49 @@ func (whisper *Whisper) runMessageLoop(p *Peer, rw p2p.MsgReadWriter) error {
 			// therefore might not satisfy the PoW, expiry and other requirements.
 			// these messages are only accepted from the trusted peer.
 			if p.trusted {
-				var envelope Envelope
-				if err := packet.Decode(&envelope); err != nil {
+				var (
+					envelope  *Envelope
+					envelopes []*Envelope
+					err       error
+				)
+
+				// Duplicate the stream because we will attempt to read twice from it
+				// in case the first read fails.
+				var buf bytes.Buffer
+				packet.Payload = io.TeeReader(packet.Payload, &buf)
```

Comment:
can you instead read whole payload into byte slice and then try decoding into different values? i think code will be easier to digest that way

Reply:
Yes, I can totally do that. I was trying to avoid reading the whole data but I am not sure if it makes any sense here. I did not measure that.

Reply:
On the other hand, it's required to read all data to parse them. I will change it then.